messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/jms/JMSService.java

```diff
@@ -164,7 +167,25 @@ public void deActivate() {
                     // and *not* during AS7 service container shutdown or reload (AS7-6840 / AS7-6881)
                     if (activeMQActivationController != null) {
                         if (!activeMQActivationController.getState().in(STOPPING, REMOVED)) {
+                            // [WFLY-4597] When Artemis is deactivated during failover, we block until its
+                            // activation controller is REMOVED before giving back control to Artemis.
+                            // This allow to properly stop any service depending on the activation controller
+                            // and avoid spurious warning messages because the resources used by the services
+                            // are stopped outside the control of the services.
+                            final CountDownLatch latch = new CountDownLatch(1);
                             activeMQActivationController.compareAndSetMode(ACTIVE, REMOVE);
+                            activeMQActivationController.addListener(new AbstractServiceListener<Void>() {
+                                @Override
+                                public void transition(ServiceController<? extends Void> controller, ServiceController.Transition transition) {
+                                    if (transition.enters(REMOVED)) {
+                                        latch.countDown();
+                                    }
+                                }
+                            });
+                            try {
+                                latch.await(5, TimeUnit.SECONDS);
+                            } catch (InterruptedException e) {
+                            }
```

Comment:
Just a note in case someone reads this someday trying to understand this change.

AFAICT in normal server/start stop this will be called as part of JMSService.stop(), but it uses StopContext.asynchronous() and passes the work off to a non-MSC thread, so this blocking behavior is ok.